rfc/0016-wasm-cache.md

```diff
@@ -0,0 +1,226 @@
+> As the output is already a compiled module, translation and code generation will be skipped and this will improve the performance of constructing a `Module` from the output of this method.
+
+Attempting to load an incompatible module will result in an error as
+stated by the [`wasmtime::Module::deserialize`](https://docs.rs/wasmtime/10.0.1/wasmtime/struct.Module.html#method.deserialize)
+documentation:
+
+> Note that this function is designed to be safe receiving output from any compiled version of wasmtime itself.
+> This means that it is safe to feed output from older versions of Wasmtime into this function, in addition to
+> newer versions of wasmtime (from the future!).
+> These inputs will deterministically and safely produce an `Err`. This function
+> only successfully accepts inputs from the same version of wasmtime, but the safety
+> guarantee only applies to externally-defined blobs of bytes, not those defined by any version of wasmtime.
+> (this means that if you cache blobs across versions of wasmtime you can be safely guaranteed that future versions of wasmtime will reject old cache entries).
+
+We can keep a cache of the previously optimized modules and attempt to reuse them.
+An incompatible module would produce an error, from which we can recover by doing
+the optimization again.
+
+It's also possible to determine if the optimized module can be reused without
+even attempting to load it. This can be done by using the
+[`wasmtime::Engine::precompile_compatibility_hash`](https://docs.rs/wasmtime/10.0.1/wasmtime/struct.Engine.html#method.precompile_compatibility_hash)
+function:
+
+> Returns a `std::hash::Hash` that can be used to check precompiled WebAssembly compatibility.
+>
+> [...] If this `Hash` matches between two `Engine`s then binaries from one are guaranteed to deserialize in the other.
+
+When optimizing an WASM module, we could also store the hash of the `Engine` that
+produced it.
+
+Later, we can use this hash to quickly find a pre-optimized WASM module that we
+could reuse.
+
+## Where to store the pre-optimized modules
+
+We could store the pre-optimized modules inside of an OCI registry.
+The Kubewarden helm chart could, when enabled via a dedicated toogle
+option, deploy an internal OCI registry.
+
+At startup time, Policy Server would query the internal registry looking for
+a pre-built WASM module. When the module is found, the Policy Server
+could download it and consume that.
+
+When not found, the Policy Server would download the vanilla WASM module, optimize
+it and then, before starting to listen for incoming requests, push the pre-built
+module to the OCI registry.
+
+Multiple Policy Server instances could be optimizing and pushing the same WASM
+module at the same time. However, the OCI registry can handle these concurrent
+push requests.
+
+## The caching OCI registry
+
+The OCI registry used for caching would not be exposed outside of the cluster.
```

Comment:
in case the cluster is airgapped, we could reuse the OCI registry inside of the airgap.